### ISSUE TITLE
Feature request: reverse_padded_sequence

### ISSUE THREAD

**Issue by rdipietro** [pytorch/pytorch#1794]
Hi,

There's existing discussion around a more extendable RNN API, which as of yet hasn't moved forward. Given that, and that `PackedSequence`s exist, would a `reverse_padded_sequence` function be useful, or do you just see it as being unnecessary clutter?

In particular it'd take in a Variable of padded sequences along with sequence lengths (and a `batch_first` option) and produce a Variable of padded sequences that are reversed according to their lengths, mimicking `reverse_sequence` in TensorFlow.

My thought is that this option would be a useful alternative to `PackedSequence`s in the case of bidirectional RNNs. (`PackedSequence`s seem to be primarily motivated by cuDNN, whereas in a research setting they might introduce more bookkeeping than they're worth.)

**Comment by rdipietro** [comment]
Possible implementation:

```
def reverse_padded_sequence(inputs, lengths, batch_first=False):
  """ Reverse sequences according to their lengths.

  Args:
    inputs: A Variable with shape `[T, B, ...]` if `batch_first`
      is False or `[B, T, ...]` if `batch_first` is True.
    lengths: A list of ints.
    batch_first: A boolean.

  Returns:
    A Variable with the same shape as `inputs` but with the
    sequences reversed.
  """
  if batch_first:
    inputs = inputs.transpose(0, 1)
  if inputs.size(1) != len(lengths):
    raise ValueError('inputs incompatible with lengths.')
  reversed_inputs = Variable(inputs.data.clone())
  for i, length in enumerate(lengths):
    time_ind = torch.LongTensor(list(reversed(range(length))))
    reversed_inputs[:length, i] = inputs[:, i][time_ind]
  if batch_first:
    reversed_inputs = reversed_inputs.transpose(0, 1)
  return reversed_inputs
```

**Comment by apaszke** [comment]
I'm ok with adding that, but I can't see a use case for it. Do you have anything specific in mind?

**Comment by rdipietro** [comment]
My use case would be bidirectional RNNs without using `PackedSequence`s or their surrounding processing.

So for example consider
- Inputs and targets at every time step
- A custom RNN cell, without options like `num_layers` and `bidirectional` (which can be handled later)
- A bidirectional RNN with multiple layers created using this cell

Here we could do something like
- obtain `inputs` from some other module (which we may want to backprop through)
- compute forward outputs over all inputs (even padding / unused values)
- compute backward outputs over all inputs via
  - reverse inputs with `reverse_padded_sequence`
  - compute forward
  - reverse outputs with `reverse_padded_sequence`
- concat forward and reverse outputs
- (repeat this whole process for however many layers we'd like)
- compute final targets
- compute the loss using only valid outputs (via indexing or masking)

**Comment by jekbradbury** [comment]
I think the overall complexity here is similar or perhaps worse than using PackedSequences, though?

**Comment by rdipietro** [comment]
Computation wise or code wise? Computation wise I do expect it to be worse or the same (depending on sequence length distribution, batch size, hidden size), but code wise I think this alternative is simpler. I haven't gone through the current boilerplate code in detail, but my guess is that all of `pack_padded_sequence`, `pad_packed_sequence`, `VariableRecurrent`, and `VariableRecurrentReverse` are specifically there for handling `PackedSequence`s? The flow above requires none of this code. Thoughts?

**Comment by apaszke** [comment]
Yeah that makes sense. Let's add it

**Comment by windweller** [comment]
Any update on this issue? It does seem like the function has been written but somehow did not end up in the actual codebase!

**Comment by rdipietro** [comment]
@apaszke @soumith

**Comment by zou3519** [comment]
This sounds reasonable. I can help review a patch that includes this.

**Comment by zwlanpishu** [comment]
> Any update on this issue? It does seem like the function has been written but somehow did not end up in the actual codebase!
> 
> For anyone looking to just use this function directly, here's a link: https://github.com/rdipietro/pytorch/blob/4c00324affb8c6d53d4362e321ea0e99ede6cfde/torch/nn/utils/rnn.py



> @apaszke @soumith

Thanks, I am also looking for a similar function to reverse the padded sequence according to their lenth. Is this implementation supported by auto-grad?

**Comment by rdipietro** [comment]
Yes

**Comment by rdipietro** [comment]
However keep in mind that this was written long ago, before it became way out of sync with newer versions of PyTorch. You might have to poke around a bit to get it to work with newer versions of PyTorch (I haven't had this need; sorry)

**Comment by speedcell4** [comment]
[`torchrua`](https://github.com/speedcell4/torchrua) provides a function, `reverse_packed_sequence`, to reverse `PackedSequence`, you can `pack_padded_sequence` your sequences first and then `pad_packed_sequence` them back. The `reverse_packed_sequence` implementation is very efficient.

```python
import torch
from torch.nn.utils.rnn import pad_sequence

from torchrua import reverse_packed_sequence, pad_packed_sequence, pack_padded_sequence

pad = pad_sequence([
    torch.arange(5) + 1,
    torch.arange(2) + 1,
    torch.arange(3) + 1,
], batch_first=True)
lengths = torch.tensor([5, 2, 3])

print(pad)
# tensor([[1, 2, 3, 4, 5],
#         [1, 2, 0, 0, 0],
#         [1, 2, 3, 0, 0]])

pack = pack_padded_sequence(pad, lengths, batch_first=True, enforce_sorted=False)
reversed_pack = reverse_packed_sequence(pack)
reversed_pad, _ = pad_packed_sequence(reversed_pack, batch_first=True)

print(reversed_pad)
# tensor([[5, 4, 3, 2, 1],
#         [2, 1, 0, 0, 0],
#         [3, 2, 1, 0, 0]])
```

**Comment by angusturner** [comment]
Agree that this would be quite useful.

My use-case is that I want to compute an auto-regressive function (non-RNN based), over the forward and reverse directions of a padded sequences.

Ended up doing:
```
def reverse_padded_sequence(tensor: Tensor, lengths: Tensor):
    """
    :param tensor: (B, max_length, *)
    :param lengths: (B,)
    :return:
    """
    out = torch.zeros_like(tensor)
    for i in range(tensor.shape[0]):
         out[i, :lengths[i]] = tensor[i, :lengths[i]].flip(dims=[0])
    return out
```

**Comment by zou3519** [comment]
Bumping to hi-pri for user activity. Will let the team make a final call on priority.

**Comment by LysSanzMoreta** [comment]
In case someone is interested, I made something similar to the above but using `map`. And it does not assume that the paddings are symbolized as 0

```python
class ReverseSequence(object):
    """Reverses sequence prior to feeding them to torch.nn.utils.rnn.pack_padded_sequence
    [A,T,R,0,0] -> [R,T,A,0,0] """
    def __init__(self,sequences,seqs_lens):
        """
        :param sequences: [N,L,feat_dim]
        :param seqs_lens: [N,]
        """
        self.sequences = sequences
        self.seqs_lens = seqs_lens.int()
    def run(self):
        reversed_list = list(map(lambda seq,seq_len: self.reverse(seq, seq_len), self.sequences,self.seqs_lens))
        return torch.concatenate(reversed_list,dim=0)

    def reverse(self,seq, seq_len):
        seq_reversed = seq
        seq_reversed[:seq_len] = seq[:seq_len].flip(dims=[0])
        return seq_reversed[None,:]
```

Please review it again before usage
